test/unit/org/apache/cassandra/db/guardrails/GuardrailPageSizeTest.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Test
+    public void testSelectStatementAgainstPageSize() throws Throwable
+    {
+        assertPagingValid("SELECT * FROM %s", 3);
+        assertPagingValid("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD);
+
+        assertPagingWarns("SELECT * FROM %s", 6, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), 6, PAGE_SIZE_WARN_THRESHOLD));
+        assertPagingWarns("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), PAGE_SIZE_ABORT_THRESHOLD, PAGE_SIZE_WARN_THRESHOLD));
+
+        assertPagingAborts("SELECT * FROM %s", 11, format("Aborting %s table query, page size %s exceeds abort threshold of %s.", currentTable(), 11, PAGE_SIZE_ABORT_THRESHOLD));
+    }
+
+    @Test
+    public void testIgnoreInternalClients() throws Throwable
+    {
+        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD);
+        assertPagingIgnored("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD);
+    }
+
+    private void assertPagingValid(String query, int pageSize) throws Throwable
+    {
+        assertValid(() -> executeWithPaging(userClientState, query, pageSize));
+    }
+
+    private void assertPagingIgnored(String query, int pageSize) throws Throwable
+    {
+        assertValid(() -> executeWithPaging(superClientState, query, pageSize));
+        assertValid(() -> executeWithPaging(systemClientState, query, pageSize));
+    }
+
+    private void assertPagingWarns(String query, int pageSize, String message) throws Throwable
+    {
+        assertWarns(() -> executeWithPaging(userClientState, query, pageSize), message);
+    }
+
+    private void assertPagingAborts(String query, int pageSize, String message) throws Throwable
+    {
+        assertAborts(() -> executeWithPaging(userClientState, query, pageSize), message);
+    }
+
+    private void executeWithPaging(ClientState state, String query, int pageSize)
+    {
+        QueryState queryState = new QueryState(state);
+
+        String formattedQuery = formatQuery(query);
+        CQLStatement statement = QueryProcessor.parseStatement(formattedQuery, queryState.getClientState());
+        statement.validate(state);
+
+        QueryOptions options = QueryOptions.create(ConsistencyLevel.ONE,
+                                                   Collections.emptyList(),
+                                                   false,
+                                                   pageSize,
+                                                   null,
+                                                   null,
+                                                   ProtocolVersion.CURRENT,
+                                                   KEYSPACE);
+
+        statement.executeLocally(queryState, options);
+    }
+
+    //not used by page-size guardrail tests.
+    protected long currentValue()
+    {
+        return 0;
```

Comment:
Nit: We are supposed not to call this because it doesn't make sense for this particular guardrail, so I would throw an exception so we quickly detect an accidental call, wdyt?
```suggestion
        throw new UnsupportedOperationException();
```

Reply:
Done :)